solidity/test/relay/Relay.test.ts

```diff
@@ -0,0 +1,1260 @@
+          const res = await relay.validateChain(proofHeaders)
+          expect(res[0]).to.be.above(0)
+        })
+
+        it("should accept valid header chains (1 before, 3 after)", async () => {
+          const proofHeaders = concatenateHexStrings(headerHex.slice(8, 12))
+          const res = await relay.validateChain(proofHeaders)
+          expect(res[0]).to.be.above(0)
+        })
+      })
+
+      context("in the new epoch", () => {
+        it("should accept valid header chains", async () => {
+          const proofHeaders = concatenateHexStrings(headerHex.slice(9, 13))
+          const res = await relay.validateChain(proofHeaders)
+          expect(res[0]).to.be.above(0)
+        })
+      })
+    })
+
+    context("with chain reorgs", () => {
+      const { postRetargetChain } = reorgHeaders
+      const reorgHex = postRetargetChain.map((header) => header.hex)
+
+      const reorgGenesis = postRetargetChain[0]
+
+      before(async () => {
+        await createSnapshot()
+        await relay
+          .connect(governance)
+          .genesis(reorgGenesis.hex, reorgGenesis.height, 8)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      context("valid chains", () => {
+        it("should be accepted", async () => {
+          const proofHeaders = concatenateHexStrings(reorgHex)
+          const res = await relay.validateChain(proofHeaders)
+          expect(res[0]).to.be.above(0)
+        })
+      })
+
+      context("invalid chains", () => {
+        it("should be rejected", async () => {
+          const pre = concatenateHexStrings(reorgHex.slice(0, 6))
+          const orphan = reorgHeaders.orphan_437478.hex
+          const post = reorgHex[7]
+          const proofHeaders = concatenateHexStrings([pre, orphan, post])
+          await expect(relay.validateChain(proofHeaders)).to.be.revertedWith(
+            "Invalid chain"
+          )
+        })
+      })
+    })
+
+    context("gas costs", () => {
+      const { chain } = headers
+      const headerHex = chain.map((header) => header.hex)
+      const retargetHeaders = concatenateHexStrings(headerHex.slice(5, 13))
+
+      before(async () => {
+        await createSnapshot()
+        await relay.connect(governance).genesis(genesisHeader, genesisHeight, 4)
+        await relay.connect(thirdParty).retarget(retargetHeaders)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      context("with proof length 6", () => {
+        it("should accept valid header chains", async () => {
+          const proofHeaders = concatenateHexStrings(headerHex.slice(5, 11))
+          await relay.connect(governance).setProofLength(6)
+          const tx = await relay.validateChainGasReport(proofHeaders)
+          const txr = await tx.wait()
+
+          expect(txr.status).to.equal(1)
+        })
+      })
+
+      context("with proof length 18", () => {
```

Comment:
Non-blocking this PR but how hard would it be to add a unit test with 50+50 mainnet retarget? This would allow us to assess mainnet gas costs and see if there are no other surprises.

Reply:
Not too hard, I just need to collect the header data for it

Reply:
Let's do it but in a separate PR.